libraries/eosiolib/contracts/eosio/key_value_singleton.hpp

```diff
@@ -0,0 +1,129 @@
+#include "../../core/eosio/datastream.hpp"
+#include "../../core/eosio/name.hpp"
+
+#include <eosio/to_key.hpp>
+
+namespace detail {
+   constexpr inline size_t max_stack_buffer_size = 512;
+}
+
+namespace eosio {
+   namespace internal_use_do_not_use {
+      extern "C" {
+         __attribute__((eosio_wasm_import))
+         int64_t kv_erase(uint64_t db, uint64_t contract, const char* key, uint32_t key_size);
+
+         __attribute__((eosio_wasm_import))
+         int64_t kv_set(uint64_t db, uint64_t contract, const char* key, uint32_t key_size, const char* value, uint32_t value_size);
+
+         __attribute__((eosio_wasm_import))
+         bool kv_get(uint64_t db, uint64_t contract, const char* key, uint32_t key_size, uint32_t& value_size);
+
+         __attribute__((eosio_wasm_import))
+         uint32_t kv_get_data(uint64_t db, uint32_t offset, char* data, uint32_t data_size);
+      }
+   }
+
+   template <typename T, eosio::name::raw SingletonName, eosio::name::raw DbName = "eosio.kvram"_n>
+   class kv_singleton {
+   public:
+      kv_singleton(eosio::name contract_name) : contract_name{contract_name} {
+         key = make_prefix(singleton_name);
+      }
+
+      ~kv_singleton() {
+         if (is_dirty) {
+            write_back();
+         }
+      }
+
+      const T& get() {
+         if (!is_cached) {
+            uint32_t copy_size;
+            uint32_t value_size;
+
+            auto success = internal_use_do_not_use::kv_get(db_name, contract_name.value, key.data(), key.size(), value_size);
+
+            eosio::check(success, "tried to get a singleton that does not exist");
+
+            void* val_buffer = value_size > detail::max_stack_buffer_size ? malloc(value_size) : alloca(value_size);
+            copy_size = internal_use_do_not_use::kv_get_data(db_name, 0, (char*)val_buffer, value_size);
+
+            T val;
+            deserialize(val, val_buffer, copy_size);
+
+            value() = val;
+            is_cached = true;
+
+            raw_original.resize(copy_size);
+            memcpy(raw_original.data(), val_buffer, copy_size);
+
+            if (value_size > detail::max_stack_buffer_size) {
+               free(val_buffer);
+            }
+         }
+
+         return value();
+      }
+
+      void set(const T& val) {
+         value() = val;
+         is_dirty = true;
+         is_cached = true;
+      }
+
+      void erase() {
+         internal_use_do_not_use::kv_erase(db_name, contract_name.value, key.data(), key.size());
+         is_cached = false;
+         is_dirty = false;
+      }
+
+      void write_back() {
```

Comment:
What if I call `write_back` before calling `get` or `set` or after calling `erase`

Reply:
Added a `is_dirty` check that should handle this case.